File: test/testRepo/src/ts/data_model.ts
/**
 * Data model interfaces and classes for the test repository.
 * This file contains multiple interfaces and classes with methods to test the code parsing functionality.
 */

/**
 * Base model interface for all data models
 */
export interface IBaseModel {
  id?: string;
  toDict(): Record<string, any>;
}

/**
 * Base model implementation
 */
export abstract class BaseModel implements IBaseModel {
  id?: string;
  
  constructor(id?: string) {
    this.id = id;
  }
  
  /**
   * Convert the model to a dictionary
   */
  toDict(): Record<string, any> {
    return {
      id: this.id
    };
  }
  
  /**
   * Create a model instance from a dictionary
   */
  static fromDict(data: Record<string, any>): BaseModel {
    throw new Error('Method must be implemented by subclass');
  }
}

/**
 * User model representing a system user
 */
export interface IUser extends IBaseModel {
  name?: string;
  email?: string;
  validate(): boolean;
}

/**
 * User model implementation
 */
export class User extends BaseModel implements IUser {
  name?: string;
  email?: string;
  
  constructor(id?: string, name?: string, email?: string) {
    super(id);
    this.name = name;
    this.email = email;
  }
  
  /**
   * Convert the user to a dictionary
   */
  toDict(): Record<string, any> {
    return {
      ...super.toDict(),
      name: this.name,
      email: this.email
    };
  }
  
  /**
   * Create a user instance from a dictionary
   */
  static fromDict(data: Record<string, any>): User {
    return new User(
      data.id,
      data.name,
      data.email
    );
  }
  
  /**
   * Validate the user data
   */
  validate(): boolean {
    if (!this.email || !this.email.includes('@')) {
      throw new Error('Invalid email address');
    }
    return true;
  }
}

/**
 * Repository model representing a code repository
 */
export interface IRepository extends IBaseModel {
  name?: string;
  url?: string;
  owner?: IUser;
  commits: ICommit[];
  addCommit(commit: ICommit): void;
}

/**
 * Commit interface representing a code commit
 */
export interface ICommit extends IBaseModel {
  message: string;
  timestamp: number;
  author: IUser;
}

/**
 * Repository model implementation
 */
export class Repository extends BaseModel implements IRepository {
  name?: string;
  url?: string;
  owner?: User;
  commits: Commit[] = [];
  
  constructor(id?: string, name?: string, url?: string, owner?: User) {
    super(id);
    this.name = name;
    this.url = url;
    this.owner = owner;
  }
  
  /**
   * Convert the repository to a dictionary
   */
  toDict(): Record<string, any> {
    return {
      ...super.toDict(),
      name: this.name,
      url: this.url,
      owner: this.owner?.toDict(),
      commits: this.commits.map(commit => commit.toDict())
    };
  }
  
  /**
   * Add a commit to the repository
   */
  addCommit(commit: Commit): void {
    this.commits.push(commit);
  }
  
  /**
   * Create a repository instance from a dictionary
   */
  static fromDict(data: Record<string, any>, users: Record<string, User> = {}): Repository {
    const repo = new Repository(
      data.id,
      data.name,
      data.url
    );
    
    // Set owner if available
    const ownerData = data.owner;
    if (ownerData && ownerData.id && users[ownerData.id]) {
      repo.owner = users[ownerData.id];
    }
    
    return repo;
  }
}

/**
 * Commit model implementation
 */
export class Commit extends BaseModel implements ICommit {
  message: string;
  timestamp: number;
  author: User;
  
  constructor(id: string, message: string, timestamp: number, author: User) {
    super(id);
    this.message = message;
    this.timestamp = timestamp;
    this.author = author;
  }
  
  /**
   * Convert the commit to a dictionary
   */
  toDict(): Record<string, any> {
    return {
      ...super.toDict(),
      message: this.message,
      timestamp: this.timestamp,
      author: this.author.toDict()
    };
  }
  
  /**
   * Create a commit instance from a dictionary
   */
  static fromDict(data: Record<string, any>, users: Record<string, User> = {}): Commit {
    const authorData = data.author;
    let author: User;
    
    if (authorData && authorData.id && users[authorData.id]) {
      author = users[authorData.id];
    } else {
      author = new User(authorData?.id, authorData?.name, authorData?.email);
    }
    
    return new Commit(
      data.id,
      data.message,
      data.timestamp,
      author
    );
  }
}

/**
 * Create a new user with the given name and email
 */
export function createUser(name: string, email: string): User {
  const user = new User(undefined, name, email);
  user.id = generateId();
  return user;
}

/**
 * Generate a unique ID
 */
export function generateId(): string {
  return Math.random().toString(36).substring(2, 15) + 
         Math.random().toString(36).substring(2, 15);
}

File: test/testRepo/src/py/data_model.py
"""
Data model classes for the test repository.
This file contains multiple classes with methods to test the code parsing functionality.
"""

class BaseModel:
    """Base class for all data models."""
    
    def __init__(self, id=None):
        """Initialize the base model with an optional ID."""
        self.id = id
    
    def to_dict(self):
        """Convert the model to a dictionary."""
        return {"id": self.id}
    
    @classmethod
    def from_dict(cls, data):
        """Create a model instance from a dictionary."""
        return cls(id=data.get("id"))


class User(BaseModel):
    """User model representing a system user."""
    
    def __init__(self, id=None, name=None, email=None):
        """Initialize a user with optional ID, name, and email."""
        super().__init__(id)
        self.name = name
        self.email = email
    
    def to_dict(self):
        """Convert the user to a dictionary."""
        data = super().to_dict()
        data.update({
            "name": self.name,
            "email": self.email
        })
        return data
    
    @classmethod
    def from_dict(cls, data):
        """Create a user instance from a dictionary."""
        return cls(
            id=data.get("id"),
            name=data.get("name"),
            email=data.get("email")
        )
    
    def validate(self):
        """Validate the user data."""
        if not self.email or "@" not in self.email:
            raise ValueError("Invalid email address")
        return True


class Repository(BaseModel):
    """Repository model representing a code repository."""
    
    def __init__(self, id=None, name=None, url=None, owner=None):
        """Initialize a repository with optional ID, name, URL, and owner."""
        super().__init__(id)
        self.name = name
        self.url = url
        self.owner = owner
        self.commits = []
    
    def to_dict(self):
        """Convert the repository to a dictionary."""
        data = super().to_dict()
        data.update({
            "name": self.name,
            "url": self.url,
            "owner": self.owner.to_dict() if self.owner else None,
            "commits": [commit.to_dict() for commit in self.commits]
        })
        return data
    
    def add_commit(self, commit):
        """Add a commit to the repository."""
        self.commits.append(commit)
    
    @classmethod
    def from_dict(cls, data, users=None):
        """Create a repository instance from a dictionary."""
        users = users or {}
        repo = cls(
            id=data.get("id"),
            name=data.get("name"),
            url=data.get("url")
        )
        
        # Set owner if available
        owner_data = data.get("owner")
        if owner_data and owner_data.get("id") in users:
            repo.owner = users[owner_data.get("id")]
        
        return repo


# Standalone function outside of any class
def create_user(name, email):
    """Create a new user with the given name and email."""
    user = User(name=name, email=email)
    user.id = generate_id()
    return user


def generate_id():
    """Generate a unique ID."""
    import uuid
    return str(uuid.uuid4())
